// @fedify/fedify/dist/mod.cjs

          const { Temporal } = require("@js-temporal/polyfill");
          const { URLPattern } = require("urlpattern-polyfill");
        
const require_transformers = require('./transformers-CoBS-oFG.cjs');
require('./compat-DmDDELst.cjs');
const require_docloader = require('./docloader-CNsCwkxG.cjs');
const require_actor = require('./actor-DOYpLAOc.cjs');
const require_middleware = require('./middleware-xSPzx3UI.cjs');
const require_lookup = require('./lookup-4na242fe.cjs');
const require_key = require('./key-Cdysd1b9.cjs');
const require_http = require('./http-CcPpEG46.cjs');
const require_proof = require('./proof-tEhbsByx.cjs');
const require_federation = require('./federation-C15gGOh1.cjs');
const require_types = require('./types-oz5uR2vt.cjs');
const require_authdocloader = require('./authdocloader-B1zJPoS6.cjs');
const require_vocab = require('./vocab-i1kbdwbr.cjs');
require('./nodeinfo-Co9lJrWl.cjs');
require('./runtime-C58AJWSv.cjs');
require('./sig-ByHXzqUi.cjs');
require('./webfinger-BjOEdFPs.cjs');

exports.Accept = require_actor.Accept;
exports.Activity = require_actor.Activity;
exports.Add = require_actor.Add;
exports.Announce = require_actor.Announce;
exports.Application = require_actor.Application;
exports.Arrive = require_actor.Arrive;
exports.Article = require_actor.Article;
exports.Audio = require_actor.Audio;
exports.Block = require_actor.Block;
exports.ChatMessage = require_actor.ChatMessage;
exports.Collection = require_actor.Collection;
exports.CollectionPage = require_actor.CollectionPage;
exports.Create = require_actor.Create;
exports.CryptographicKey = require_actor.CryptographicKey;
exports.DataIntegrityProof = require_actor.DataIntegrityProof;
exports.Delete = require_actor.Delete;
exports.DidService = require_actor.DidService;
exports.Dislike = require_actor.Dislike;
exports.Document = require_actor.Document;
exports.Emoji = require_actor.Emoji;
exports.EmojiReact = require_actor.EmojiReact;
exports.Endpoints = require_actor.Endpoints;
exports.Event = require_actor.Event;
exports.Export = require_actor.Export;
exports.FetchError = require_docloader.FetchError;
exports.Flag = require_actor.Flag;
exports.Follow = require_actor.Follow;
exports.Group = require_actor.Group;
exports.Hashtag = require_actor.Hashtag;
exports.Ignore = require_actor.Ignore;
exports.Image = require_actor.Image;
exports.InProcessMessageQueue = require_federation.InProcessMessageQueue;
exports.IntransitiveActivity = require_actor.IntransitiveActivity;
exports.Invite = require_actor.Invite;
exports.Join = require_actor.Join;
exports.LanguageString = require_actor.LanguageString;
exports.Leave = require_actor.Leave;
exports.Like = require_actor.Like;
exports.Link = require_actor.Link;
exports.Listen = require_actor.Listen;
exports.MemoryKvStore = require_federation.MemoryKvStore;
exports.Mention = require_actor.Mention;
exports.Move = require_actor.Move;
exports.Multikey = require_actor.Multikey;
exports.Note = require_actor.Note;
exports.Object = require_actor.Object$1;
exports.Offer = require_actor.Offer;
exports.OrderedCollection = require_actor.OrderedCollection;
exports.OrderedCollectionPage = require_actor.OrderedCollectionPage;
exports.Organization = require_actor.Organization;
exports.PUBLIC_COLLECTION = require_vocab.PUBLIC_COLLECTION;
exports.Page = require_actor.Page;
exports.ParallelMessageQueue = require_federation.ParallelMessageQueue;
exports.Person = require_actor.Person;
exports.Place = require_actor.Place;
exports.Profile = require_actor.Profile;
exports.PropertyValue = require_actor.PropertyValue;
exports.Question = require_actor.Question;
exports.Read = require_actor.Read;
exports.Reject = require_actor.Reject;
exports.Relationship = require_actor.Relationship;
exports.Remove = require_actor.Remove;
exports.Router = require_middleware.Router;
exports.RouterError = require_middleware.RouterError;
exports.Service = require_actor.Service;
exports.Source = require_actor.Source;
exports.TentativeAccept = require_actor.TentativeAccept;
exports.TentativeReject = require_actor.TentativeReject;
exports.Tombstone = require_actor.Tombstone;
exports.Travel = require_actor.Travel;
exports.Undo = require_actor.Undo;
exports.Update = require_actor.Update;
exports.Video = require_actor.Video;
exports.View = require_actor.View;
exports.actorDehydrator = require_transformers.actorDehydrator;
exports.attachSignature = require_proof.attachSignature;
exports.autoIdAssigner = require_transformers.autoIdAssigner;
exports.buildCollectionSynchronizationHeader = require_middleware.buildCollectionSynchronizationHeader;
exports.createExponentialBackoffPolicy = require_middleware.createExponentialBackoffPolicy;
exports.createFederation = require_middleware.createFederation;
exports.createFederationBuilder = require_middleware.createFederationBuilder;
exports.createProof = require_proof.createProof;
exports.createSignature = require_proof.createSignature;
exports.detachSignature = require_proof.detachSignature;
exports.digest = require_middleware.digest;
exports.doesActorOwnKey = require_proof.doesActorOwnKey;
exports.exportJwk = require_key.exportJwk;
exports.exportMultibaseKey = require_actor.exportMultibaseKey;
exports.exportSpki = require_actor.exportSpki;
exports.fetchDocumentLoader = require_docloader.fetchDocumentLoader;
exports.fetchKey = require_key.fetchKey;
exports.formatSemVer = require_types.formatSemVer;
exports.generateCryptoKeyPair = require_key.generateCryptoKeyPair;
exports.getActorClassByTypeName = require_actor.getActorClassByTypeName;
exports.getActorHandle = require_actor.getActorHandle;
exports.getActorTypeName = require_actor.getActorTypeName;
exports.getAuthenticatedDocumentLoader = require_authdocloader.getAuthenticatedDocumentLoader;
exports.getDefaultActivityTransformers = require_transformers.getDefaultActivityTransformers;
exports.getDocumentLoader = require_docloader.getDocumentLoader;
exports.getKeyOwner = require_proof.getKeyOwner;
exports.getNodeInfo = require_types.getNodeInfo;
exports.getTypeId = require_actor.getTypeId;
exports.getUserAgent = require_docloader.getUserAgent;
exports.importJwk = require_key.importJwk;
exports.importMultibaseKey = require_actor.importMultibaseKey;
exports.importPem = require_actor.importPem;
exports.importPkcs1 = require_actor.importPkcs1;
exports.importSpki = require_actor.importSpki;
exports.isActor = require_actor.isActor;
exports.isFediverseHandle = require_vocab.isFediverseHandle;
exports.kvCache = require_docloader.kvCache;
exports.lookupObject = require_vocab.lookupObject;
exports.lookupWebFinger = require_lookup.lookupWebFinger;
exports.nodeInfoToJson = require_types.nodeInfoToJson;
exports.normalizeActorHandle = require_actor.normalizeActorHandle;
exports.parseFediverseHandle = require_vocab.parseFediverseHandle;
exports.parseNodeInfo = require_types.parseNodeInfo;
exports.parseSemVer = require_types.parseSemVer;
exports.respondWithObject = require_middleware.respondWithObject;
exports.respondWithObjectIfAcceptable = require_middleware.respondWithObjectIfAcceptable;
exports.signJsonLd = require_proof.signJsonLd;
exports.signObject = require_proof.signObject;
exports.signRequest = require_http.signRequest;
exports.toAcctUrl = require_vocab.toAcctUrl;
exports.traverseCollection = require_vocab.traverseCollection;
exports.verifyJsonLd = require_proof.verifyJsonLd;
exports.verifyObject = require_proof.verifyObject;
exports.verifyProof = require_proof.verifyProof;
exports.verifyRequest = require_http.verifyRequest;
exports.verifySignature = require_proof.verifySignature;
