// @fedify/fedify/dist/federation/mod.cjs

          const { Temporal } = require("@js-temporal/polyfill");
          const { URLPattern } = require("urlpattern-polyfill");
        
require('../transformers-CoBS-oFG.cjs');
require('../docloader-CNsCwkxG.cjs');
require('../actor-DOYpLAOc.cjs');
const require_middleware = require('../middleware-xSPzx3UI.cjs');
require('../lookup-4na242fe.cjs');
require('../key-Cdysd1b9.cjs');
require('../http-CcPpEG46.cjs');
require('../proof-tEhbsByx.cjs');
const require_federation = require('../federation-C15gGOh1.cjs');
require('../types-oz5uR2vt.cjs');
require('../authdocloader-B1zJPoS6.cjs');
require('../vocab-i1kbdwbr.cjs');

exports.InProcessMessageQueue = require_federation.InProcessMessageQueue;
exports.MemoryKvStore = require_federation.MemoryKvStore;
exports.ParallelMessageQueue = require_federation.ParallelMessageQueue;
exports.Router = require_middleware.Router;
exports.RouterError = require_middleware.RouterError;
exports.buildCollectionSynchronizationHeader = require_middleware.buildCollectionSynchronizationHeader;
exports.createExponentialBackoffPolicy = require_middleware.createExponentialBackoffPolicy;
exports.createFederation = require_middleware.createFederation;
exports.createFederationBuilder = require_middleware.createFederationBuilder;
exports.digest = require_middleware.digest;
exports.respondWithObject = require_middleware.respondWithObject;
exports.respondWithObjectIfAcceptable = require_middleware.respondWithObjectIfAcceptable;
