// @fedify/fedify/dist/federation/mod.js

          import { Temporal } from "@js-temporal/polyfill";
          import { URLPattern } from "urlpattern-polyfill";
        
import "../transformers-BFT6d7J5.js";
import "../docloader-BRNFZnTb.js";
import "../actor-jvzet8bO.js";
import { Router, RouterError, buildCollectionSynchronizationHeader, createExponentialBackoffPolicy, createFederation, createFederationBuilder, digest, respondWithObject, respondWithObjectIfAcceptable } from "../middleware-BXUQTcBK.js";
import "../lookup-CKEj5FWO.js";
import "../key-gldrk_5Z.js";
import "../http-C3_5amY1.js";
import "../proof-Bx8lj3_m.js";
import { InProcessMessageQueue, MemoryKvStore, ParallelMessageQueue } from "../federation-K-xsO03l.js";
import "../types-BsFw6XaG.js";
import "../authdocloader-Ddqc8DWD.js";
import "../vocab-qsTMTRUj.js";

export { InProcessMessageQueue, MemoryKvStore, ParallelMessageQueue, Router, RouterError, buildCollectionSynchronizationHeader, createExponentialBackoffPolicy, createFederation, createFederationBuilder, digest, respondWithObject, respondWithObjectIfAcceptable };
