// @fedify/fedify/dist/vocab/mod.cjs

          const { Temporal } = require("@js-temporal/polyfill");
          const { URLPattern } = require("urlpattern-polyfill");
        
require('../docloader-CNsCwkxG.cjs');
const require_actor = require('../actor-DOYpLAOc.cjs');
require('../lookup-4na242fe.cjs');
const require_vocab = require('../vocab-i1kbdwbr.cjs');

exports.Accept = require_actor.Accept;
exports.Activity = require_actor.Activity;
exports.Add = require_actor.Add;
exports.Announce = require_actor.Announce;
exports.Application = require_actor.Application;
exports.Arrive = require_actor.Arrive;
exports.Article = require_actor.Article;
exports.Audio = require_actor.Audio;
exports.Block = require_actor.Block;
exports.ChatMessage = require_actor.ChatMessage;
exports.Collection = require_actor.Collection;
exports.CollectionPage = require_actor.CollectionPage;
exports.Create = require_actor.Create;
exports.CryptographicKey = require_actor.CryptographicKey;
exports.DataIntegrityProof = require_actor.DataIntegrityProof;
exports.Delete = require_actor.Delete;
exports.DidService = require_actor.DidService;
exports.Dislike = require_actor.Dislike;
exports.Document = require_actor.Document;
exports.Emoji = require_actor.Emoji;
exports.EmojiReact = require_actor.EmojiReact;
exports.Endpoints = require_actor.Endpoints;
exports.Event = require_actor.Event;
exports.Export = require_actor.Export;
exports.Flag = require_actor.Flag;
exports.Follow = require_actor.Follow;
exports.Group = require_actor.Group;
exports.Hashtag = require_actor.Hashtag;
exports.Ignore = require_actor.Ignore;
exports.Image = require_actor.Image;
exports.IntransitiveActivity = require_actor.IntransitiveActivity;
exports.Invite = require_actor.Invite;
exports.Join = require_actor.Join;
exports.Leave = require_actor.Leave;
exports.Like = require_actor.Like;
exports.Link = require_actor.Link;
exports.Listen = require_actor.Listen;
exports.Mention = require_actor.Mention;
exports.Move = require_actor.Move;
exports.Multikey = require_actor.Multikey;
exports.Note = require_actor.Note;
exports.Object = require_actor.Object$1;
exports.Offer = require_actor.Offer;
exports.OrderedCollection = require_actor.OrderedCollection;
exports.OrderedCollectionPage = require_actor.OrderedCollectionPage;
exports.Organization = require_actor.Organization;
exports.PUBLIC_COLLECTION = require_vocab.PUBLIC_COLLECTION;
exports.Page = require_actor.Page;
exports.Person = require_actor.Person;
exports.Place = require_actor.Place;
exports.Profile = require_actor.Profile;
exports.PropertyValue = require_actor.PropertyValue;
exports.Question = require_actor.Question;
exports.Read = require_actor.Read;
exports.Reject = require_actor.Reject;
exports.Relationship = require_actor.Relationship;
exports.Remove = require_actor.Remove;
exports.Service = require_actor.Service;
exports.Source = require_actor.Source;
exports.TentativeAccept = require_actor.TentativeAccept;
exports.TentativeReject = require_actor.TentativeReject;
exports.Tombstone = require_actor.Tombstone;
exports.Travel = require_actor.Travel;
exports.Undo = require_actor.Undo;
exports.Update = require_actor.Update;
exports.Video = require_actor.Video;
exports.View = require_actor.View;
exports.getActorClassByTypeName = require_actor.getActorClassByTypeName;
exports.getActorHandle = require_actor.getActorHandle;
exports.getActorTypeName = require_actor.getActorTypeName;
exports.getTypeId = require_actor.getTypeId;
exports.isActor = require_actor.isActor;
exports.isFediverseHandle = require_vocab.isFediverseHandle;
exports.lookupObject = require_vocab.lookupObject;
exports.normalizeActorHandle = require_actor.normalizeActorHandle;
exports.parseFediverseHandle = require_vocab.parseFediverseHandle;
exports.toAcctUrl = require_vocab.toAcctUrl;
exports.traverseCollection = require_vocab.traverseCollection;
