// @fedify/fedify/dist/x/hono.cjs

          const { Temporal } = require("@js-temporal/polyfill");
          const { URLPattern } = require("urlpattern-polyfill");
        
const require_chunk = require('../chunk-DqRYRqnO.cjs');
const __logtape_logtape = require_chunk.__toESM(require("@logtape/logtape"));

//#region src/x/hono.ts
/**
* Create a Hono middleware to integrate with the {@link Federation} object.
*
* @deprecated This function has been moved to `@fedify/hono` package.
*             Import `federation` from `@fedify/hono` instead.
*             This function will be removed in Fedify v2.0.
*
* @template TContextData A type of the context data for the {@link Federation}
*                         object.
* @template THonoContext A type of the Hono context.
* @param federation A {@link Federation} object to integrate with Hono.
* @param contextDataFactory A function to create a context data for the
*                           {@link Federation} object.
* @returns A Hono middleware.
*/
function federation(federation$1, contextDataFactory) {
	const logger = (0, __logtape_logtape.getLogger)([
		"fedify",
		"federation",
		"hono"
	]);
	logger.warn("The `@fedify/fedify/x/hono` module is deprecated; use `federation` from `@fedify/hono` package instead.");
	return async (ctx, next) => {
		let contextData = contextDataFactory(ctx);
		if (contextData instanceof Promise) contextData = await contextData;
		return await federation$1.fetch(ctx.req.raw, {
			contextData,
			...integrateFetchOptions(ctx, next)
		});
	};
}
function integrateFetchOptions(ctx, next) {
	return {
		async onNotFound(_req) {
			await next();
			return ctx.res;
		},
		async onNotAcceptable(_req) {
			await next();
			if (ctx.res.status !== 404) return ctx.res;
			return new Response("Not acceptable", {
				status: 406,
				headers: {
					"Content-Type": "text/plain",
					Vary: "Accept"
				}
			});
		}
	};
}

//#endregion
exports.federation = federation;
