// @fedify/fedify/dist/x/sveltekit.cjs

          const { Temporal } = require("@js-temporal/polyfill");
          const { URLPattern } = require("urlpattern-polyfill");
        
const require_chunk = require('../chunk-DqRYRqnO.cjs');
const __logtape_logtape = require_chunk.__toESM(require("@logtape/logtape"));

//#region src/x/sveltekit.ts
/**
* Create a SvelteKit hook handler to integrate with the {@link Federation}
* object.
*
* @deprecated This function has been moved to `@fedify/sveltekit` package.
*             Import `fedifyHook` from `@fedify/sveltekit` instead.
*             This function will be removed in Fedify v2.0.
*
* @example hooks.server.ts
* ``` typescript
* import { federation } from "./federation"; // Import the `Federation` object
*
* export const handle = fedifyHook(federation, () => undefined);
* ```
*
* @template TContextData A type of the context data for the {@link Federation}
*                         object.
* @param federation A {@link Federation} object to integrate with SvelteKit.
* @param createContextData A function to create a context data for the
*                          {@link Federation} object.
* @returns A SvelteKit hook handler.
* @since 1.3.0
*/
function fedifyHook(federation, createContextData) {
	const logger = (0, __logtape_logtape.getLogger)([
		"fedify",
		"federation",
		"sveltekit"
	]);
	logger.warn("The `@fedify/fedify/x/sveltekit` module is deprecated; use `fedifyHook` from `@fedify/sveltekit` package instead.");
	return async ({ event, resolve }) => {
		return await federation.fetch(event.request, {
			contextData: await createContextData(event),
			...integrateFetchOptions({
				event,
				resolve
			})
		});
	};
}
function integrateFetchOptions({ event, resolve }) {
	return {
		async onNotFound() {
			return await resolve(event);
		},
		async onNotAcceptable() {
			const res = await resolve(event);
			if (res.status !== 404) return res;
			return new Response("Not acceptable", {
				status: 406,
				headers: {
					"Content-Type": "text/plain",
					Vary: "Accept"
				}
			});
		}
	};
}

//#endregion
exports.fedifyHook = fedifyHook;
