// @fedify/fedify/dist/sig/mod.cjs

          const { Temporal } = require("@js-temporal/polyfill");
          const { URLPattern } = require("urlpattern-polyfill");
        
require('../docloader-CNsCwkxG.cjs');
require('../actor-DOYpLAOc.cjs');
require('../lookup-4na242fe.cjs');
const require_key = require('../key-Cdysd1b9.cjs');
const require_http = require('../http-CcPpEG46.cjs');
const require_proof = require('../proof-tEhbsByx.cjs');
require('../sig-ByHXzqUi.cjs');

exports.attachSignature = require_proof.attachSignature;
exports.createProof = require_proof.createProof;
exports.createSignature = require_proof.createSignature;
exports.detachSignature = require_proof.detachSignature;
exports.doesActorOwnKey = require_proof.doesActorOwnKey;
exports.exportJwk = require_key.exportJwk;
exports.fetchKey = require_key.fetchKey;
exports.generateCryptoKeyPair = require_key.generateCryptoKeyPair;
exports.getKeyOwner = require_proof.getKeyOwner;
exports.importJwk = require_key.importJwk;
exports.signJsonLd = require_proof.signJsonLd;
exports.signObject = require_proof.signObject;
exports.signRequest = require_http.signRequest;
exports.verifyJsonLd = require_proof.verifyJsonLd;
exports.verifyObject = require_proof.verifyObject;
exports.verifyProof = require_proof.verifyProof;
exports.verifyRequest = require_http.verifyRequest;
exports.verifySignature = require_proof.verifySignature;
